test/test_pipeline_schedule.py

```diff
@@ -79,225 +21,58 @@ def __init__(
-            device_type="cuda", mesh_shape=(int(os.environ["WORLD_SIZE"]),)
-        )
+        self.w1 = nn.Linear(dim, hidden_dim, bias=False)
+        self.w2 = nn.Linear(hidden_dim, out_dim, bias=False)
+        self.relu = nn.ReLU()
 
-        rank = device_mesh.get_rank()
-        rank_print(f"using {device_mesh=}")
+    def forward(self, x):
+        x = self.w1(x)
+        x = self.w2(x)
+        x = self.relu(x)
+        return x
 
-    if not device_mesh:
-        rank = kwargs["rank"]
 
-    init_process_group = bool(device_mesh is not None)
+# Tests defined below
+##########################
 
-    setup(local_rank, init_process_group=init_process_group)
 
-    logger.info(
-        f"====== World Rank {rank}, Local Rank {local_rank}, World Size {world_size}, Device {device} main ======"
-    )
+# python -m unittest test_pipeline_schedule.TestPipelineSchedule.<test>
+#               or
+# pytest test_pipeline_schedule.py -vsk <test>
+class TestPipelineSchedule(MultiProcessTestCase):
+    @property
+    def world_size(self) -> int:
+        return 4
 
-    rank_print(f"My KWARGS are {kwargs}")
+    def setUp(self):
+        super().setUp()
+        # starts world_size processes
+        self._spawn_processes()
 
-    input_dim = 4000
-    hidden_dim = 8000
-    output_dim = 4000
-
-    module_list = torch.nn.ModuleList(
-        modules=[
-            MLP(input_dim, hidden_dim, output_dim) for i in range(world_size)
-        ]
-    )
-    microbatch_size = 8
-    global_batch_size = 64
-    assert global_batch_size % microbatch_size == 0
-    n_microbatches = int(global_batch_size / microbatch_size)
-    n_pp = world_size
-
-    x = torch.randn([microbatch_size, input_dim]).to("meta")
-    unused = torch.ones((1, 1), device="meta")
-    input_args = (x, unused)
-
-    stage_model = PipelineStageV2Impl(
-        module_list[rank],
-        rank,
-        world_size,
-        rank,
-        world_size,
-        input_args,
-        device,
-    )
-    stage_model.init_p2p_neighbors()
-
-    stage_model_looped = [
+    def test_pipeline_stage_init(self):
+        model_meta = MLP(dim=8, hidden_dim=4, out_dim=4)
+        input_meta = torch.rand((2, 8), device="meta")
         PipelineStageV2Impl(
-            module_list[rank],
-            stage_id=(world_size * i) + rank,
-            num_stages=world_size * world_size,
-            rank=rank,
-            world_size=world_size,
-            input_args=input_args,
-            device=device,
-        )
-        for i in range(world_size)
-    ]
-    x_cuda_empty = torch.empty_like(x, device="cuda")
-
-    microbatches = []
-    for i in range(n_microbatches):
-        microbatches.append(
-            (torch.randn_like(x_cuda_empty), torch.ones((1, 1), device="cuda"))
+            model_meta,
+            stage_id=0,
+            num_stages=1,
+            rank=0,
+            world_size=1,
+            input_args=[input_meta],
+            device="cuda",
```

Comment:
thinking out loud, if we use .cpu we can run testing in CI today. if we have to do a lot of stuff to make cpu/gloo work nicely with pipelinestage, this might be less appealing but.. worth a try?

Reply:
yeah, that makes sense. That would make it a lot easier to test, everything should be able to "just work" with gloo

Reply:
hmm. even batch_isend_irecv?

Reply:
oh yeah thats a good point. in gloo, batch isend_irecv is implemented multiples send/recv operations rather than a single operations as with nccl. So schedules like gpipe / looped bfs should still work, looped DFS we may see that hang due to conflicting ops that we saw a while back.

Reply:
if the op is "literally" supported by gloo, then you can use cpu to do unit tests of pipelinestage, etc.  But maybe not worth trying to get integration/e2e testing working if unreliable. we should get gpu machines.